mutex.go

```diff
@@ -0,0 +1,108 @@
+// Copyright 2016 Canonical Ltd.
+// Licensed under the LGPLv3, see LICENCE file for details.
+
+package mutex
+
+import (
+	"regexp"
+	"time"
+
+	"github.com/juju/errors"
+)
+
+var (
+	validName = regexp.MustCompile("^[a-zA-Z][a-zA-Z0-9-]*$")
+)
+
+// Releaser defines the Release method that is the only thing that can be done
+// to a acquired mutex.
+type Releaser interface {
+	// Release makes the mutex available for others and invalidates the instance.
+	// Subsequent calls to Release will error.
+	Release() error
+}
+
+// Acquirer defines the standard method to acquire a mutex.
+type Acquirer interface {
+	Acquire() (Releaser, error)
+}
+
+// Clock provides an interface for dealing with clocks.
+type Clock interface {
+	// After waits for the duration to elapse and then sends the
+	// current time on the returned channel.
+	After(time.Duration) <-chan time.Time
+}
+
+// Spec defines the name of the mutex and behaviour of the Acquire function.
+type Spec struct {
+	// Name is required, and must start with a letter and contain at most
+	// 40 letters, numbers or dashes.
+	Name string
+	// Clock must be provided and is exposed for testing purposes.
+	Clock Clock
+	// Delay defines how often to check for lock acquisition. The implementation
+	// calls all fail fast, so acquisition requires polling.
+	Delay time.Duration
+	// Timeout allows the caller to specify how long to wait.
+	Timeout time.Duration
+	// Cancel if signalled will cause the Acquire method to return with ErrCancelled.
+	Cancel <-chan struct{}
+}
+
+// Acquire implements Acquirer.
```

Comment:
I agree that there should only be one way, either Spec.Acquire or Acquire. Sticking with Spec.Acquire has the benefit of easier mocking via an interface when testing something that takes one of these.